### DISCUSSION TITLE
Extending request processing

### DISCUSSION THREAD

**Discussion by atasmrk007** [CoreWCF/CoreWCF#809] [Q&A]
I have a question about extending the request and response processing. When request is received, I would like to log the request received and start a stopwatch. When request is processed I want to stop the stopwatch and log the response with the note of how long the processing took.

I would very much like to avoid doing that within an "OperationContract". With WebApi I can do that with message filters, extending the pipeline, whatever. How would I achieve something like that with CoreWCF?

**Comment by g7ed6e** [comment]
You can use a `IDispatchMessageInspector` to hook in the the pipeline.
```c#
public object AfterReceiveRequest(ref System.ServiceModel.Channels.Message request, IClientChannel channel, InstanceContext instanceContext)
{
  Stopwatch stopWatch = new Stopwatch();
  stopWatch.Start();
  
  // return the stopWatch as correlationState
  return stopWatch;
}

public void BeforeSendReply(ref System.ServiceModel.Channels.Message reply, object correlationState)
{
  // retrieve the StopWatch from correlationState
  Stopwatch stopWatch = (Stopwatch)correlationState;
  stopWatch.Stop();
  
  // Get the elapsed time as a TimeSpan value.
  TimeSpan ts = stopWatch.Elapsed;
}
```
Please see https://docs.microsoft.com/en-us/dotnet/api/system.servicemodel.dispatcher.idispatchmessageinspector?view=dotnet-plat-ext-6.0 to add it to the `DispatchRuntime` using a `IServiceBehavior`.

Regarding logging the request, WCF had message logging builtin https://docs.microsoft.com/en-us/dotnet/framework/wcf/diagnostics/configuring-message-logging?redirectedfrom=MSDN but i do not know the status of the port of this feature with CoreWCF

**Reply by mconnew** [reply]
We don't have message logging. It depends on a couple of other features we haven't added yet, Trace logging and/or ETW logging. It also was only configurable in the .config file, there was no way to configure it in code. We'll need to come up with a new way to do this.  
The IDispatchMessageInspector way is the transport agnostic way of doing things. It does run after authentication and the entire request has been read from the socket though (for buffered transfer mode). If you want to do it the asp.net core way, you could do it with some middleware. You can do something like this:
```c#
app.Use(next => async context =>
{
    Stopwatch stopWatch = new Stopwatch();
    stopWatch.Start();

    await next(context);

    stopWatch.Stop();
    // Get the elapsed time as a TimeSpan value.
    TimeSpan ts = stopWatch.Elapsed;
});
app.UseServiceModel(builder => { .... });
```

**Reply by atasmrk007** [reply]
Thank you for your sugestions. I went with a middleware solution, trying to be .NET compliant. I really want to log every request and response so I thought it should not be a problem to add those, right? Well, I encountered two issues.

1. Streams being destroyed when read complicates the matter somewhat, but there are a lot of resources on how to work around this issue. You can set request stream to have buffering enabled and that solves the stream being null after read for request. Response is a whole another beast. Basically you have to swap response stream with a stream that can be "rewound", the rest of the pipeline will work with your stream. When you are in control again, you do whatever with the stream that got filled and then copy stream content into original stream. [stack overflow question](https://stackoverflow.com/questions/43403941/how-to-read-asp-net-core-response-body) Working like that I can log the web service landing page response no problem. If I am trying to log SOAP response however I am unable to move back and forth on the stream. I get the "System.ObjectDisposedException: Cannot access a closed Stream." exception. That leads me to believe that CoreWCF is flushing/closing the stream when it should not.

2. I wanted to take care of logging exceptions. I implemented [.net suggested solution](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/error-handling?view=aspnetcore-6.0#exception-handler-lambda) and put a test throw into one of my methods. But the added code is never invoked. So again I suspect that CoreWCF is doing something that messes with ASP.NET.

Would I have to open Issues for the problems I have?

**Reply by bpashkovskyi** [reply]
@atasmrk007 were you able to fix issues with flushed response Streams? I have the same

**Reply by alexvazquez** [reply]
> You can use a `IDispatchMessageInspector` to hook in the the pipeline.
> 
> ```cs
> public object AfterReceiveRequest(ref System.ServiceModel.Channels.Message request, IClientChannel channel, InstanceContext instanceContext)
> {
>   Stopwatch stopWatch = new Stopwatch();
>   stopWatch.Start();
>   
>   // return the stopWatch as correlationState
>   return stopWatch;
> }
> 
> public void BeforeSendReply(ref System.ServiceModel.Channels.Message reply, object correlationState)
> {
>   // retrieve the StopWatch from correlationState
>   Stopwatch stopWatch = (Stopwatch)correlationState;
>   stopWatch.Stop();
>   
>   // Get the elapsed time as a TimeSpan value.
>   TimeSpan ts = stopWatch.Elapsed;
> }
> ```
> 
> Please see https://docs.microsoft.com/en-us/dotnet/api/system.servicemodel.dispatcher.idispatchmessageinspector?view=dotnet-plat-ext-6.0 to add it to the `DispatchRuntime` using a `IServiceBehavior`.
> 
> Regarding logging the request, WCF had message logging builtin https://docs.microsoft.com/en-us/dotnet/framework/wcf/diagnostics/configuring-message-logging?redirectedfrom=MSDN but i do not know the status of the port of this feature with CoreWCF

Hi Do you have a code example on how to use the `IDispatchMessageInspector `in the project configuration? Thanks a lot

**Reply by birojnayak** [reply]
tagging the issue #790 as we have plan for open telemetry to give these information's out of the box.

**Comment by ThisBytes5** [comment]
> 1. If I am trying to log SOAP response however I am unable to move back and forth on the stream. I get the "System.ObjectDisposedException: Cannot access a closed Stream." exception. That leads me to believe that CoreWCF is flushing/closing the stream when it should not.

I know it's been almost a year, but any chance you figured this out? I'm running into the same issue now.
